sdk/core/azure-core-amqp/inc/azure/core/amqp/connection.hpp

```diff
@@ -0,0 +1,167 @@
+// Copyright(c) Microsoft Corporation.All rights reserved.
+// SPDX - License - Identifier : MIT
+
+#pragma once
+
+#include "common/async_operation_queue.hpp"
+#include "connection_string_credential.hpp"
+#include "models/amqp_value.hpp"
+#include "session.hpp"
+#include <azure/core/credentials/credentials.hpp>
+#include <chrono>
+#include <memory>
+#include <string>
+
+namespace Azure { namespace Core { namespace _internal { namespace Amqp {
+
+  namespace _detail {
+    class ConnectionImpl;
+  }
+
+  class Session;
+
+  struct ConnectionOptions
+  {
+    //    std::chrono::seconds IdleTimeout{std::chrono::seconds(1)};
+    uint32_t MaxFrameSize{512};
+    uint16_t MaxSessions{65535};
+
+    Azure::Core::Amqp::Models::Value Properties;
+
+    std::string SASLType; // Not a string - fill in later.
+
+    std::chrono::seconds Timeout{0};
+
+    /** Enable tracing from the uAMQP stack.
+     */
+    bool EnableTrace{false};
+
+    /** Defines the ID of the container for this connection. If empty, a unique 128 bit value will
+     * be used.
+     */
+    std::string ContainerId;
+
+    std::string HostName;
+    uint16_t Port{5671}; // Assume TLS port by default.
+
+    // Default transport to be used. Normally only needed for socket listeners which need to specify
+    // the listening socket characteristics.
+    std::shared_ptr<Azure::Core::_internal::Amqp::Network::Transport> Transport{};
+
+    // Optional SASL plain credentials.
+    std::shared_ptr<Azure::Core::_internal::Amqp::SaslPlainConnectionStringCredential>
+        SaslCredentials{};
+  };
+
+  class Error;
+
+  enum class ConnectionState
+  {
+    Start,
+    HeaderReceived,
+    HeaderSent,
+    HeaderExchanged,
+    OpenPipe,
+    OcPipe,
+    OpenReceived,
+    OpenSent,
+    ClosePipe,
+    Opened,
+    CloseReceived,
+    CloseSent,
+    Discarding,
+    End,
+    Error
+  };
+
+  class Connection;
+
+  struct ConnectionEvents
+  {
+    virtual void OnEndpointFrameReceived(
+        Connection const& connection,
+        Azure::Core::Amqp::Models::Value value,
```

Comment:
In this case, it should be, because it's the AMQP implementation providing a value. In most of the other cases, when a value is accepted as an "in" parameter, it shouldn't because we want to allow implicit conversions to Value on input.